[JSC] Expose error messages from ShadowRealms to outer realm

caitp opened the pull request:
#### ea7efa2f7f885f2dc5d0dea65d185fb9974cbf58
<pre>
[JSC] Expose error messages from ShadowRealms to outer realm
<a href="https://bugs.webkit.org/show_bug.cgi?id=249324">https://bugs.webkit.org/show_bug.cgi?id=249324</a>

Reviewed by Yusuke Suzuki.

This adds stuff from a yet unmerged spec change from <a href="https://github.com/tc39/proposal-shadowrealm/pull/382.">https://github.com/tc39/proposal-shadowrealm/pull/382.</a>

This particular implementation only concerns itself with converting a
primitive exception to a String and using that as the TypeError message,
or else accessing the own &quot;message&quot; property of an object and using
that, which is sufficient to inherit messages from a native Error.

A more robust approach could attempt to load &quot;message&quot; from the
prototype chain, and could incorporate some limited source information
from the source error into the new error message.

Canonical link: <a href="https://commits.webkit.org/258438@main">https://commits.webkit.org/258438@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34f7f8eb91afcfb06f490e838661c2d8d986f988

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101974 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11119 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35046 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111307 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12087 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2035 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94382 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109061 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107755 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/94382 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/23996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94382 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/92362 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4702 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/25437 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/88537 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2309 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4792 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/1877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29448 "Found 1 new JSC stress test failure: wasm.yaml/wasm/lowExecutableMemory/exports-oom.js.default-wasm (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10868 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/44925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/91451 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5789 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6541 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20442 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

caitp commented on Source/JavaScriptCore/runtime/Error.cpp:
```diff
@@ -255,6 +255,33 @@ JSObject* addErrorInfo(VM& vm, JSObject* error, int line, const SourceCode& sour
+        }
+    }
+
+    scope.releaseAssertNoException();
```
this _may_ want to check for termination exceptions too, not sure.

Constellation replied:
Yes, this is not correct when termination exception happens.
Let's use `RETURN_IF_EXCEPTION(scope, {})`

Constellation replied:
And drop this.

Constellation commented on Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:
```diff
@@ -105,22 +105,25 @@ JSC_DEFINE_HOST_FUNCTION(evalInRealm, (JSGlobalObject* globalObject, CallFrame*
             throwVMError(globalObject, scope, createSyntaxError(globalObject, syntaxErrorMessage));
         } else
-            throwVMError(globalObject, scope, createTypeError(globalObject, "Error encountered during evaluation"_s));
+            throwVMError(globalObject, scope, createTypeErrorCopy(globalObject, error));
```
Let's add `RETURN_IF_EXCEPTION` after `createTypeErrorCopy` call.

Constellation commented on Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:
```diff
@@ -105,22 +105,25 @@ JSC_DEFINE_HOST_FUNCTION(evalInRealm, (JSGlobalObject* globalObject, CallFrame*
+        JSValue error = exception->value();
         scope.clearException();
-        return throwVMError(globalObject, scope, createTypeError(globalObject, "Error encountered during evaluation"_s));
+        return throwVMError(globalObject, scope, createTypeErrorCopy(globalObject, error));
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/Error.cpp:
```diff
@@ -255,6 +255,33 @@ JSObject* addErrorInfo(VM& vm, JSObject* error, int line, const SourceCode& sour
+                scope.assertNoException();
+                if (slot.isValue()) {
+                    JSValue message = slot.getValue(globalObject, vm.propertyNames->message);
+                    scope.assertNoException();
```
This is not correct when termination exception happens. So, use

```
RETURN_IF_EXCEPTION(scope, {})
```

caitp replied:
Done 👍

Constellation commented on Source/JavaScriptCore/runtime/Error.cpp:
```diff
@@ -255,6 +255,34 @@ JSObject* addErrorInfo(VM& vm, JSObject* error, int line, const SourceCode& sour
+    String errorString = "Error encountered during evaluation"_s;
+
+    if (error.isPrimitive())
+        errorString = error.toWTFString(globalObject);
```
We need exception check here to suppress warnings.

caitp replied:
Would it be terrible to add a version of toWTFString that doesn't throw or invoke user code?

caitp replied:
Regardless, done

Constellation commented on Source/JavaScriptCore/runtime/Error.cpp:
```diff
@@ -255,6 +255,34 @@ JSObject* addErrorInfo(VM& vm, JSObject* error, int line, const SourceCode& sour
+                if (slot.isValue()) {
+                    JSValue message = slot.getValue(globalObject, vm.propertyNames->message);
+                    RETURN_IF_EXCEPTION(scope, { });
+                    errorString = message.toWTFString(globalObject);
```
Ditto. Add exception check

Constellation commented on Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:
```diff
@@ -105,22 +105,30 @@ JSC_DEFINE_HOST_FUNCTION(evalInRealm, (JSGlobalObject* globalObject, CallFrame*
+        } else {
+            auto typeError = createTypeErrorCopy(globalObject, error);
+            RETURN_IF_EXCEPTION(scope, { });
+            throwVMError(globalObject, scope, typeError);
```
Need to return.

Constellation commented on Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp:
```diff
@@ -105,22 +105,30 @@ JSC_DEFINE_HOST_FUNCTION(evalInRealm, (JSGlobalObject* globalObject, CallFrame*
-            const String syntaxErrorMessage = error->sanitizedMessageString(globalObject);
+            const String syntaxErrorMessage = errorInstance->sanitizedMessageString(globalObject);
             RETURN_IF_EXCEPTION(scope, { });
             throwVMError(globalObject, scope, createSyntaxError(globalObject, syntaxErrorMessage));
```
Need to return.

Constellation reviewed (commented):
More comments.

Constellation reviewed (commented):
Nice, looks good.

caitp committed:
[JSC] Expose error messages from ShadowRealms to outer realm
https://bugs.webkit.org/show_bug.cgi?id=249324

Reviewed by Yusuke Suzuki.

This adds stuff from a yet unmerged spec change from https://github.com/tc39/proposal-shadowrealm/pull/382.

This particular implementation only concerns itself with converting a
primitive exception to a String and using that as the TypeError message,
or else accessing the own "message" property of an object and using
that, which is sufficient to inherit messages from a native Error.

A more robust approach could attempt to load "message" from the
prototype chain, and could incorporate some limited source information
from the source error into the new error message.

Canonical link: https://commits.webkit.org/258438@main
